MatrixHelperGUI/LinearMatrixGUI.cs

```diff
@@ -100,50 +97,58 @@ private void btnAddRows_Click(object sender, EventArgs e)
                 _matrix.SumRows(rowIndex1, rowIndex2, rowReplace);
                 UpdateLog("R" + cboResultantRow.Text + ": R" + cboRow1.Text + " + R" + cboRow2.Text);
             }
+            PushMatrixToStack(_matrix);
             PrintMatrixToLabel(_matrix, lblMatrixDisplay);
         }
         private void btnMultiplyRow_Click(object sender, EventArgs e)
         {
-            _matrix.MultiplyRow(Convert.ToInt32(cboRow1.Text), Convert.ToDouble(txtFactor.Text));
-            UpdateLog("R" + cboRow1.Text + ": " + txtFactor.Text + cboRow1.Text);
-            PrintMatrixToLabel(_matrix, lblMatrixDisplay);
-        }
-        private void cboRow1_SelectedIndexChanged(object sender, EventArgs e)
-        {
-            UpdateResultantRowComboBox();
-        }
-        private void cboRow2_SelectedIndexChanged(object sender, EventArgs e)
-        {
-            UpdateResultantRowComboBox();
-        }
-        private void btnMultiplyRows_Click(object sender, EventArgs e)
-        {
             double factor;
             string factorText = txtFactor.Text.Replace('\\', '/'); //Just in case the wrong slash is used
             if (factorText.Contains("/"))
             {
                 double n = Convert.ToDouble(factorText.Substring(0, factorText.IndexOf('/')));
                 double d = Convert.ToDouble(factorText.Substring(factorText.IndexOf('/') + 1));
                 factor = n / d;
-            } else
+            }
+            else
             {
                 factor = Convert.ToDouble(factorText);
             }
             _matrix.MultiplyRow(Convert.ToInt32(cboRow1.Text) - 1, factor);
-            PrintMatrixToLabel(_matrix, lblMatrixDisplay);
+            PushMatrixToStack(_matrix);
             UpdateLog("R" + cboRow1.Text + ": " + factorText + "R" + cboRow1.Text);
+            PrintMatrixToLabel(_matrix, lblMatrixDisplay);
+        }
+        private void cboRow1_SelectedIndexChanged(object sender, EventArgs e)
+        {
+            UpdateResultantRowComboBox();
+        }
+        private void cboRow2_SelectedIndexChanged(object sender, EventArgs e)
+        {
+            UpdateResultantRowComboBox();
         }
         private void btnSwapRows_Click(object sender, EventArgs e)
         {
             _matrix.SwitchRows(Convert.ToInt32(cboRow1.Text) - 1, Convert.ToInt32(cboRow2.Text) - 1);
             PrintMatrixToLabel(_matrix, lblMatrixDisplay);
             UpdateLog("R" + cboRow1.Text + " <---> R" + cboRow2.Text);
+            PushMatrixToStack(_matrix);
         }
         private void btnUndo_Click(object sender, EventArgs e)
         {
-            _matrix.Undo();
-            PrintMatrixToLabel(_matrix, lblMatrixDisplay);
-            UpdateLog("Undo last action.");
+            if(_stack.Count > 1)
+            {
+                _matrix =_stack.Pop();
```

Comment:
This line doesn't need to be _matrix = _stack.Pop();  it doesn't hurt anything to have it like that, but it should just be _stack.Pop();  I'll fix it later.  Or should I do a whole commit just to fix this one thing?  What is best practice?